Updates to AlpakaCore README and tests

makortel opened the pull request:
#### PR description:

This PR makes some updates to the `AlpakaCore` README file, mainly to reflect https://github.com/cms-sw/cmssw/pull/41019 and to clarify a few things. It also extends the testing to cover the cases where the Alpaka backend is explicitly specified via the `ProcessAcceleratorAlpaka`, and moves to use `!serial` backend specification for the `DataFormats/PortableTestObjects`.

Resolves https://github.com/makortel/framework/issues/517

#### PR validation:

Unit tests run both with and without GPU.

makortel committed:
Update documentation

makortel committed:
Use not-serial for specifying the Alpaka backends in PortableTestObjects

makortel committed:
Add tests for setting the default backend in ProcessAcceleratorAlpaka

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

- DataFormats/PortableTestObjects (**heterogeneous**)
- HeterogeneousCore/AlpakaCore (**heterogeneous**)
- HeterogeneousCore/AlpakaTest (**heterogeneous**)


@fwyzard, @makortel, @cmsbuild can you please review it and eventually sign? Thanks.
@missirol, @rovere this is something you requested to watch as well.
@sextonkennedy, @rappoccio, @antoniovilela you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
enable gpu

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-GPU 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35216/summary.html
**COMMIT**: 942aa9b45fd61c74407b40748299e4ed9ea90db7
**CMSSW**: CMSSW_13_3_X_2023-10-16-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43033/35216/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-GPU

- [12434.587](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35216/runTheMatrixGPU-results/12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation)```12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation/step2_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation.log```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3357378
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-16-2300+0d1c0c/59267/triggerResults): no differences found

makortel commented:
Rebased (although was not not strictly necessary), and updated the README for https://github.com/cms-sw/cmssw/pull/43041

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43033/37240

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaTest/test/testAlpakaModules_cfg.py  modified in PR(s): #41564

cmsbuild commented:
Pull request #43033 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: RelVals-GPU 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35242/summary.html
**COMMIT**: ac125753386de417f4a2003d530e08f057cb151c
**CMSSW**: CMSSW_13_3_X_2023-10-17-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43033/35242/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-GPU

- [12434.587](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35242/runTheMatrixGPU-results/12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation)```12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation/step2_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation.log```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 9
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3357369
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-17-2300+0d1c0c/59282/triggerResults): no differences found

makortel committed:
Use better names for the PortableCollection and templated data format models in the comments

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43033/37259

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaTest/interface/alpaka/AlpakaESTestData.h  modified in PR(s): #43041
  - File  HeterogeneousCore/AlpakaTest/test/testAlpakaModules_cfg.py  modified in PR(s): #41564

cmsbuild commented:
Pull request #43033 was updated. @makortel, @cmsbuild, @fwyzard can you please check and sign again.

makortel commented:
@cmsbuild, please test

fwyzard commented on HeterogeneousCore/AlpakaCore/README.md:
```diff
@@ -42,13 +42,17 @@ Data formats, for both Event and EventSetup, should be placed following their us
   * For EventSetup data products [the registration macro `TYPELOOKUP_DATA_REG`](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToRegisterESData) should be placed in `Package/SubPackage/src/ES_<type name>.cc`.
 * The device-side data formats are defined in `Package/SubPackage/interface/alpaka/` directory
   * The device-side data format classes should be either templated over the device type, or defined in the `ALPAKA_ACCELERATOR_NAMESPACE` namespace.
+  * For host backends (`serial`), the "device-side" data format class must be the same as the aforementioned host0only data format class
```
```suggestion
  * For host backends (`serial`), the "device-side" data format class must be the same as the aforementioned host-only data format class
```

fwyzard commented on HeterogeneousCore/AlpakaCore/README.md:
```diff
@@ -352,14 +358,13 @@ There are a few different options for using Alpaka-based modules in the CMSSW co
 ## Load explicitly
-# One ProcessAccelerator for each accelerator technology
-process.load("HeterogeneousCore.CUDACore.ProcessAcceleratorCUDA_cfi")
+# One ProcessAccelerators for each accelerator technology
```
```suggestion
# One ProcessAccelerator for each accelerator technology
```

fwyzard commented on HeterogeneousCore/AlpakaCore/README.md:
```diff
@@ -408,3 +413,33 @@ process.producerCPU = cms.EDProducer("ExampleAlpakaProducer@alpaka",
+
+#### Examples on explicitly setting the backend
+
+##### Per module instance
```
I think it would be clearer like
```suggestion
##### For individual modules
```

fwyzard commented on HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerA.cc:
```diff
@@ -11,9 +11,10 @@
-   * This class demonstrates and ESProducer on the data model 1 that
-   * consumes a standard host ESProduct and converts the data into
-   * PortableCollection, and implicitly transfers the data product to device
+   * This class demonstrates and ESProducer that uses the
```
```suggestion
   * This class demonstrates an ESProducer that uses the
```

fwyzard commented on HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerB.cc:
```diff
@@ -11,11 +11,12 @@
-   * Alpaka buffer that is then moved into an object of a class that
-   * is templated over the device type, and implicitly transfers the
-   * data product to device
+   * This class demonstrates and ESProducer that uses the
```
```suggestion
   * This class demonstrates an ESProducer that uses the
```

fwyzard commented on HeterogeneousCore/AlpakaTest/plugins/alpaka/TestAlpakaESProducerC.cc:
```diff
@@ -14,9 +14,10 @@
-   * This class demonstrates and ESProducer on the data model 1 that
-   * consumes a standard host ESProduct and converts the data into
-   * PortableCollection, and implicitly transfers the data product to device
+   * This class demonstrates and ESProducer that uses the
```
```suggestion
   * This class demonstrates an ESProducer that uses the
```

fwyzard commented on HeterogeneousCore/AlpakaTest/test/BuildFile.xml:
```diff
@@ -1,6 +1,3 @@
-<!-- TODO: dependence on CUDA only to make GPU IBs to run these tests -->
-<use name="cuda"/>
-
```
?

makortel replied:
Was a mistake

fwyzard reviewed (changes requested):
Mostly typos.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43033/37280

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/AlpakaTest/interface/alpaka/AlpakaESTestData.h  modified in PR(s): #43041
  - File  HeterogeneousCore/AlpakaTest/test/testAlpakaModules_cfg.py  modified in PR(s): #41564

cmsbuild commented:
-1

**Failed Tests**: RelVals-GPU 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35291/summary.html
**COMMIT**: 7b35b38782db705bdecc10e00cc9aca72141dca8
**CMSSW**: CMSSW_13_3_X_2023-10-19-1100/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43033/35291/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-GPU

- [12434.587](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35291/runTheMatrixGPU-results/12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation)```12434.587_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation/step2_TTbar_14TeV+2023_Patatrack_AllTripletsGPU_Validation.log```


## Comparison Summary



**Summary**:
 - You potentially removed 4 lines from the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3357372
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-19-1100+0d1c0c/59319/triggerResults): no differences found

makortel commented:
The GPU unit tests are run now again (thanks @fwyzard for spotting it!). GPU RelVal failure is https://github.com/cms-sw/cmssw/issues/42997.

I addressed all comments of @fwyzard .

makortel commented:
+heterogeneous

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

fwyzard commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0d1c0c/35336/summary.html
**COMMIT**: 7b35b38782db705bdecc10e00cc9aca72141dca8
**CMSSW**: CMSSW_13_3_X_2023-10-22-0000/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43033/35336/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3357400
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3357375
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 214 log files, 167 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-22-0000+0d1c0c/59354/triggerResults): no differences found


## GPU Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 3
 - DQMHistoTests: Total histograms compared: 39740
 - DQMHistoTests: Total failures: 20
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 39720
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 2 files compared)
 - Checked 8 log files, 10 edm output root files, 3 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsGPU/CMSSW_13_3_X_2023-10-22-0000+0d1c0c/59356/triggerResults): no differences found

rappoccio commented:
+1
